pkg/autoscaler/statforwarder/processor.go

```diff
@@ -17,28 +17,69 @@ limitations under the License.
 package statforwarder
 
 import (
+	"sync"
+	"time"
+
 	gorillawebsocket "github.com/gorilla/websocket"
 	"go.uber.org/zap"
 
 	"knative.dev/pkg/websocket"
 	asmetrics "knative.dev/serving/pkg/autoscaler/metrics"
 )
 
+// The timeout value for a Websocket connection to be established. If a connection via IP
+// address can not be established within this value, we assume the Pods can not be
+// accessed by IP address directly due to the network mesh.
+const establishTimeout = 500 * time.Millisecond
+
 // bucketProcessor includes the information about how to process
 // the StatMessage owned by a bucket.
 type bucketProcessor struct {
 	logger *zap.SugaredLogger
 	// The name of the bucket
 	bkt string
 	// holder is the HolderIdentity for a bucket from the Lease.
-	holder string
+	holder   string
+	svcDNS   string
+	connLock sync.RWMutex
 	// conn is the WebSocket connection to the holder pod.
 	conn *websocket.ManagedConnection
 	// `accept` is the function to process a StatMessage which doesn't need
 	// to be forwarded.
 	accept statProcessor
 }
 
+func newForwardProcessor(logger *zap.SugaredLogger, bkt, holder, podDNS, svcDNS string) *bucketProcessor {
+	// First try to connect via Pod IP address synchronously. If the connection can
+	// not be established within `establishTimeout`, we assume the pods can not be
+	// accessed by IP address. Then try to connect via Pod IP address synchronously.
+	logger.Info("Connecting to Autoscaler bucket at ", podDNS)
+	c, err := websocket.NewDurableSendingConnectionGuaranteed(podDNS, establishTimeout, logger)
+	if err != nil {
+		logger.Info("Autoscaler pods can't be accessed by IP address. Connecting to Autoscaler bucket at ", svcDNS)
+		c, _ = websocket.NewDurableSendingConnectionGuaranteed(svcDNS, establishTimeout, logger)
```

Comment:
We probably need to examine the error here: if it's timeout retry (unlikely) if it's something else — fatal probably?

Reply:
Currently it can only return one err: `ErrConnectionNotEstablished`. We need to add support in pkg repo.